compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -715,6 +711,36 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
     }
 }
 
+/// Whether we should encode MIR.
+///
+/// Return a pair, resp. for CTFE and for LLVM.
+fn should_encode_mir(tcx: TyCtxt<'_>, def_id: LocalDefId) -> (bool, bool) {
+    match tcx.def_kind(def_id) {
+        // Constructors
+        DefKind::Ctor(_, _) => {
+            let mir_opt_base = tcx.sess.opts.output_types.should_codegen()
+                || tcx.sess.opts.debugging_opts.always_encode_mir;
+            (true, mir_opt_base)
+        }
+        // Constants
+        DefKind::AnonConst | DefKind::AssocConst | DefKind::Static | DefKind::Const => {
+            (true, false)
+        }
+        // Closures and functions
+        DefKind::Generator | DefKind::Closure | DefKind::AssocFn | DefKind::Fn => {
```

Comment:
Currently optimized MIR for generators is always required. I will have to revisit the tests if this implementation manges to pass them.